### ISSUE TITLE
Update the contributing guidelines to describe the procedure for merging accepted pull requests

### ISSUE THREAD

**Issue by davidhassell** [cf-convention/cf-conventions#194]
**Title:** Update the contributing guidelines to describe the procedure for merging accepted pull requests.

**Moderator:** 

**Requirement Summary:** The CONTRIBUTING.md file does not include information on the merge procedure of accepted changes. However, understanding how and when your accepted issue makes its way into the published conventions is important to all contributors. 

**Technical Proposal Summary:** This issue aims to remedy this by describing how pull requests are merged, by whom, and and on what timescale.

**Benefits:** These changes will benefit anyone involved with updating the CF conventions.

**Status Quo:** There is confusion on how merges should be done, and even if they should be done, yet.

**Detailed Proposal:** The proposed extra text to CONTRIBUTING.md may be seen in teh following Pull Request. The new text, occurs in three sections, is:

Some best practice not wholly related to merges:
```
5. **Comments in pull requests.** These should be avoided, unless
discussing changes to the wording of a proposal that do not impact on
the agreed meaning. This is so that the scientific development of the
proposal is easily found in one place, i.e. the GitHub issue.

```
```
It is recommended that pull requests are created on a branch of a
personal fork of this repository.
```

The merge procedure:
```
## Merging of pull requests, and closing and reopening of issues

When an issue and accompanying pull request concludes with an agreed
change to the conventions, then the pull request will be merged into
the master branch after a further 3 weeks has elapsed, assuming that
no more discussion occurs. This is as described by the rules for
conventions changes at http://cfconventions.org/rules.html. If further
substantive comments are raised during this period, then the final 3
week period will restart after a new conclusion is agreed.

The pull request will be merged by a member of the CF conventions and
standard names committee.

The person who merges the pull request should also close the issue,
leaving a comment that the issue was concluded with no further comment
during the subsequent 3 weeks.

If subsequent discussion is required after the pull request has been
merged then a new issue should be raised, rather than reopening the
closed issue.

An issue may be closed without the merging of a pull request if the
change was not accepted by the community. In this case the issue may
be reopened for further discussion at a later date.
```

Note that I have written that _"The pull request will be merged by a member of the CF conventions and standard names committee"_. I make this suggestion because the committee is formally responsible for "publishing new versions of the CF Conventions as needed", and it's large enough that it is almost certain that some members will active at any given time.

Some of this content should also be reflected in the "Rules for Conventions Changes" document (http://cfconventions.org/rules.html), but it makes sense to agree this here, and then submit a change issue afterwards for the Rules document in its repository (https://github.com/cf-convention/cf-convention.github.io).

**Comment by dblodgett-usgs** [comment]
Dear @davidhassell -- Thanks for this issue and PR. I'll play moderator of this issue and PR #195. I am supportive of the language here. 

Is there any discussion or disagreement from others in the community?

**Comment by JonathanGregory** [comment]
Dear @davidhassell  and Dave @dblodgett-usgs
These changes look good and I support them. Thanks.
I think the CONTRIBUTING file should appear on the website, and I wonder whether it ought to be merged with the existing rules.html file. However that doesn't affect the content of it so is probably a subsequent issue to consider.
Jonathan

**Comment by dblodgett-usgs** [comment]
Dear @JonathanGregory, I agree. @davidhassell do you want to open an issue to follow up with @JonathanGregory suggests?

**Comment by martinjuckes** [comment]
Dear @davidhassell ,

I support the approach, but there appears to be a contradiction between "All new pull requests should be submitted to the master branch and will be merged or closed as soon as agreement has been reached" and "When an issue and accompanying pull request concludes with an agreed
change to the conventions, then the pull request will be merged into the master branch after a further 3 weeks has elapsed, assuming that no more discussion occurs". I think the latter is what we want: can the former sentence be deleted?

Making the committee members responsible for carrying out pull requests is not a nice solutions, but it may be the best approach available now. Ideally, I think we should have a group of people with the technical skills and availability to run the platform. However, the changes you propose are already a significant step forward, and I don't want to hold things up on this point.

**Comment by davidhassell** [comment]
Hello @martinjuckes 

I agree that there is a contradiction from the original and proposed text, and that the original ("All new pull requests should be submitted to the master branch and will be merged or closed as soon as agreement has been reached") can be deleted. I'll update the PR. Thanks.

I appreciate your acceptance for now, if a bit reluctantly, to specifying that the merge should come from the Committee. Expanding on my thinking, merging is not technically hard (just one click), but understanding the rules that determine if a merge should happen is a lot more involved, and should be done by someone who has accepted responsibility for making changes to CF, i.e. someone on the committee. The currently numbers 12 people (http://cfconventions.org/governance.html). Do you not think that that's a sufficient pool of people?

Thanks,
David

**Comment by davidhassell** [comment]
PR updated.
```
All new pull requests should be submitted to the master branch and will be merged or
closed as soon as agreement has been reached.
```
replaced with
```
All new pull requests should be submitted to the master branch of this repository.
```

**Comment by martinjuckes** [comment]
Hi David, 

My concern with having the Committtee responsible for executing pull requests was that it mixes up different roles. Just one click you say, but they obviously need to know where to click, and we don't appear to have supporting documentation for this kind of role. But if the committee is happy with it, I'll drop the objection.

regards,
Martin

**Comment by davidhassell** [comment]
I see your point. My view is that it is important for a well defined group to have this responsibility, and this area seems to fit in seamlessly with the existing responsibilities of the committee. The committee:

* already exists;
* is well defined (http://cfconventions.org/governance.html) and
* is _" responsible for developing changes to the CF standard"_ (http://cfconventions.org/Data/cf-documents/cf-governance/cf2_whitepaper_final.html).

Making sure that the committee knows what to do is just as easy as making sure that any other small group does, and by definition the committee has a vested interest in making sure that things progress.

As Martin says, it would be very useful to hear from some other members of the committee to see what their views are, seeing as it's they who might have this responsibility placed on their shoulders! The opinions of the governance panel are equally important.

**Comment by dblodgett-usgs** [comment]
IMHO, it is de-facto the same group. At the end of the day, I think being on the conventions committee or governance panel should give a person merge privileges on the repository. Anyone with those privileges should be able to merge a PR. After all, a PR can always be undone if there ends up being disagreement and there are plenty of people with the knowledge of how to manage the process on the committee.

**Comment by JonathanGregory** [comment]
Dear all

I think it would be OK to give merge privilege to both the committee and the panel. At present nearly all the people on the panel are also on the committee, but they could all be different, and I think helps to have a larger group. I agree with David's arguments that these groups have advantages in being well-defined and presently in existence, with defined responsibilities. The panel does not have responsibility for the detail of the convention, but it knows what it's about! None of this implies that other people could not do this job equally responsibly. If we find we need more, we can revisit this later and define a further group somehow.

It would be useful for most of these people (including me) to have simple instructions on how to do this job.

Cheers, Jonathan

**Comment by davidhassell** [comment]
Hello,

I agree that in principle to including the CONTIRUBUTING file on the website, but won't that mean that it exists in two repositories (https://github.com/cf-convention/cf-conventions and https://github.com/cf-convention/cf-convention.github.io), which means that they are likely get out of sync? Is there a way around this?

In PR #195 I have added the governance panel to the list of those who may merge (in addition to the committee): 
```
The pull request will be merged by a member of the governance panel or the CF
conventions and standard names committee.
```

With regards merge instructions, it is very simple, so could they be added to CONTRIBUTING.md? @dblodgett-usgs, which if the three types of merge would you recommend ("Create a merge commit", "Squash and merge", "Rebase and merge")? On paper, I would say the "Squash" option, for which all commits on the PR branch are squashed to  a single commit on the cf-conventions repository, as opposed to each individual PR commit appearing on the cf-conventions. What do people think? 

(P.S. I am out of the office for a couple of weeks from tomorrow)

**Comment by JonathanGregory** [comment]
Dear David

If we're going to continue to use the website repo (which we have to call
cf-convention.github.io, it turns out, don't we) and that repo is also the
one for governance, shouldn't we keep the CONTRIBUTING file *only* in the
cf-convention.github.io repo? It's more about rules of procedure than part
of the convention.

Best wishes

Jonathan

**Comment by davidhassell** [comment]
Dear Jonathan,

I see your point about moving all of this to conventions.github.io, but I think that it is also standard github  procedure to expect it in CONTRIBUTING.md. 

Seeing as we already have a CONTRIBUTING.md, could I suggest that we keep it, modifying as suggesting by this issue. This issue can then, hopefully, be resolved and closed - thereby cementing the principle that we can start using as soon as possible.

How CONTRIBUTING.md and the rules on the website interact is still a conversation that I would like to have, but that could perhaps be raised as a separate issue.

Is that OK?

All the best,
David

**Comment by erget** [comment]
TL;DR - this looks good to me.

Coming in late to the game - one item to nitpick is that the PR says

> The pull request will be merged by a member of the governance panel or the CF conventions **and** standard names committee.

Emphasis is mine.

ATM these are merged but only in practice, not per regulation, and I hope that #172 makes it a bit clearer that these are separate bodies, unless we make the explicit decision to merge them later. So that sentence will need to be revisited if we make the separation more explicit.

Also, in #172 I want to introduce the idea of motions so that the 3 weeks rule might be changed slightly - if adopted. Just a thing to keep in mind.

I also agree that the Committee should have merge privileges on the repository. It would be prudent to due a workshop on this sometime, maybe next year, to lower the barrier - once you learn how to do it it's really easy. From a procedural perspective the Committee "owns" some documents and the Panel "owns" others so that would also be an intuitive division of privileges across repositories in the organisation.

Regarding where to put the rules, I agree, the separation of the rules per se and the GitHub implementation is an issue to itself and it probably does make sense to have them as separate, cross-linked documents. I agree with @davidhassell  that this is probably a separate issue.

**Comment by JonathanGregory** [comment]
Dear David

I see - I didn't know that CONTRIBUTING.md has a conventional status in GitHub.
OK. I don't want to delay our migration to GitHub, so let's do as you say.

However, I do think we ought to resolve this as soon as we can. For example,
maybe we could make the CONTRIBUTING.md a minimal file, which provided little
more than a link to a webpage.

Best wishes

Jonathan
